texteditor.py

```diff
@@ -0,0 +1,195 @@
+# Copyright (C) 2015, Batchu Venkat Vishal
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+# Boston, MA 02111-1307, USA.
+
+import logging
+import time
+
+from gi.repository import Gtk
+
+try:
+    from sugar3.presence.wrapper import CollabWrapper
+    logging.error('USING SUGAR COLLAB WRAPPER')
+except ImportError:
+    from collabwrapper import CollabWrapper
+
+'''
+The collabtexteditor moudle provides a text editor widget 
+which can be included in any activity and then multiple 
+users can collaborate and edit together in the editor.
+'''
+
+
+class CollabTextEditor(Gtk.TextView):
+    '''
+    A CollabTextEditor widget is a adjustable text editor which
+    can be placed on an activity screen.
+
+    The `changed` signal is usually emitted when the text in the
+    editor is changed by a user.
+    The `message` signal is usually emitted when another user makes
+    changes in the text editor, so they are reflected in your editor.
+
+    The widget can be embedded in a window which can be displayed.
+    Example usage:
+        editorinstance = CollabTextEditor(self)
+        scrolled_window.add(editorinstance)
+        scrolled_window.show()
+
+    '''
+
+    def __init__(self, activity, editor_id, collab):
+        Gtk.TextView.__init__(self)
+        self._id = editor_id
+        self._callbacks_status = True
+        collab.connect('message', self.__message_cb)
+        collab.connect('buddy-joined', self.__buddy_joined_cb)
+        if not hasattr(collab, 'setup_done'):
+            collab.setup_done = True
+            collab.setup()
```

Comment:
Why do you call collab.setup?  Shouldn't that be the activities duty?

Calling collab.setup can cause bad things to happen if it called before the activity is ready, so it is important that it gets called at only the right time.